base/compiler/ssair/inlining.jl

```diff
@@ -1010,13 +1055,95 @@ function recompute_method_matches(atype, sv)
+                only_method = meth[1][3]
+            elseif only_method !== meth[1][3]
+                only_method = false
+            end
+        else
+            only_method = false
+        end
+        for match in meth::Vector{Any}
+            (metharg, methsp, method) = (match[1]::Type, match[2]::SimpleVector, match[3]::Method)
+            # TODO: This could be better
+            signature_union = Union{signature_union, metharg}
+            if !isdispatchtuple(metharg)
+                fully_covered = false
+                continue
+            end
+            case_sig = Signature(sig.f, sig.ft, sig.atypes, metharg)
+            case = analyze_method!(idx, case_sig, metharg, methsp, method,
+                stmt, sv, false, nothing, calltype)
+            if case === nothing
+                fully_covered = false
+                continue
+            elseif _any(p->p[1] === metharg, cases)
+                continue
+            end
+            push!(cases, Pair{Any,Any}(metharg, case))
+        end
+    end
+
+    too_many && return
+
+    signature_fully_covered = sig.atype <: signature_union
+    # If we're fully covered and there's only one applicable method,
+    # we inline, even if the signature is not a dispatch tuple
+    if signature_fully_covered && length(cases) == 0 && only_method isa Method
+        if length(infos) > 1
+            method = only_method
+            (metharg, methsp) = ccall(:jl_type_intersection_with_env, Any, (Any, Any),
+                sig.atype, method.sig)::SimpleVector
+        else
+            @assert length(meth) == 1
+            (metharg, methsp, method) = (meth[1][1]::Type, meth[1][2]::SimpleVector, meth[1][3]::Method)
+        end
+        fully_covered = true
+        case = analyze_method!(idx, sig, metharg, methsp, method,
+            stmt, sv, false, nothing, calltype)
+        case === nothing && return
+        push!(cases, Pair{Any,Any}(metharg, case))
+    end
+    if !signature_fully_covered
+        fully_covered = false
+    end
+
+    # If we only have one case and that case is fully covered, we may either
+    # be able to do the inlining now (for constant cases), or push it directly
+    # onto the todo list
+    if fully_covered && length(cases) == 1
+        handle_single_case!(ir, stmt, idx, cases[1][2], false, todo)
+        return
+    end
+    length(cases) == 0 && return
+    push!(todo, UnionSplit(idx, fully_covered, sig.atype, cases))
+end
+
 function assemble_inline_todo!(ir::IRCode, sv::OptimizationState)
     # todo = (inline_idx, (isva, isinvoke, na), method, spvals, inline_linetable, inline_ir, lie)
     todo = Any[]
     skip = find_throw_blocks(ir.stmts.inst, RefValue(ir))
     for idx in 1:length(ir.stmts)
         idx in skip && continue
-        r = process_simple!(ir, idx, sv.params, sv.world)
+        r = process_simple!(ir, todo, idx, sv.params, sv.world, sv)
         r === nothing && continue
 
         stmt = ir.stmts[idx][:inst]
```

Comment:
In theory, the purpose of `ir.stmts[idx]` existing as a type now is so that we don't need to pass idx, stmt, type, and so on as separate arguments in new code